### DISCUSSION TITLE
Why are so many tokens sent on every request?

### DISCUSSION THREAD

**Discussion by LL1yt** [cline/cline#366] [Feature Requests]
Today I tested extension for the first time, and within an hour, I hit the daily token limit—reaching 1 million tokens. In the stats, I see that each request is about 60k tokens, though the code so far is only about 20k characters. It feels like the entire chat history is being sent every time, and 90% of it is going to the expensive cache. Is there a way to see why so many tokens are being sent/added to the cache, and/or to configure or optimize this process?

**Comment by CiberNin** [comment]
The entire history IS being sent each time. Until you hit the context limit. That's just how LLM's work. The LLM only "remembers" what you feed back in. The cache usually ends up being way cheaper.
Your limit will increase if you load your account with more credits. (may be delayed depending on how long since you first created account)

**Reply by LL1yt** [reply]
The cache was created precisely so as not to send the same information every time. the cost of reading cache is up to 90% less [Prompt caching with Claude](https://www.anthropic.com/news/prompt-caching) That’s why I’m confused about why the entire chat history is being written to the cache with every request, especially since caching is more expensive than just sending tokens with the request. Wouldn’t it be more efficient to read the existing conversation from a cheaper cache and only add the new messages as they come in? Moreover, it's unnecessary to cache certain responses like when the LLM apologizes for a mistake and explains how it was corrected. I personally would exclude those. However, since this app is for a wider audience, it would make sense to have a setting that allows users to control what gets cached and what doesn’t.

**Comment by saoudrizwan** [comment]
> It feels like the entire chat history is being sent every time

As @CiberNin said the entire conversation history does get sent up with each request, but on Anthropic's servers only the non-cached parts have to be "re-processed". Now I have no idea what this means, probably Anthropic tokenizing and converting the prompt into a special format for the model. The fact this processing is so expensive is very interesting... anyways there's a couple of things to keep in mind here, this cache "breaks" when certain things happen, like sending an image or modifying even 1 character in your conversation history including the system prompt. So if you were to update Settings > Custom Instructions halfway through a task, it would break the cache and you would be charged for writing to the cache again. Another instance the cache will break is when you hit your 200k context window limit and the extension will truncate the first half of your conversation history to let you keep the task going (my own take on context window management to take advantage of prompt caching).
